src/session/content/gradient_bg.rs

```diff
@@ -0,0 +1,497 @@
+            let color1 = graphene::Vec3::new(0.1, 1.0, 0.5);
+            let color2 = graphene::Vec3::new(1.0, 1.0, 0.5);
+            let color3 = graphene::Vec3::new(0.1, 1.0, 1.0);
+            let color4 = graphene::Vec3::new(0.5, 0.0, 1.0);
+
+            Self {
+                shaders: Default::default(),
+                pattern: RefCell::new(pattern),
+
+                gradient_cache: TextureCache {
+                    texture: RefCell::new(None),
+                    params: RefCell::new(GradientCacheParams {
+                        size: (0.0, 0.0),
+                        colors: [color1, color2, color3, color4],
+                        phase: 0,
+                    }),
+                },
+
+                progress: Default::default(),
+                phase: Default::default(),
+
+                color1: Cell::new(color1),
+                color2: Cell::new(color2),
+                color3: Cell::new(color3),
+                color4: Cell::new(color4),
+                dark: Default::default(),
+            }
+        }
+    }
+
+    #[glib::object_subclass]
+    impl ObjectSubclass for GradientBackground {
+        const NAME: &'static str = "ComponentsGradientBackground";
+        type Type = super::GradientBackground;
+        type ParentType = gtk::Widget;
+    }
+
+    impl ObjectImpl for GradientBackground {
+        fn constructed(&self) {
+            self.parent_constructed();
+
+            let obj = self.obj();
+
+            obj.set_vexpand(true);
+            obj.set_hexpand(true);
+
+            let target = adw::CallbackAnimationTarget::new(clone!(@weak obj => move |progress| {
+                let imp = obj.imp();
+                let progress = progress as f32;
+                if progress >= 1.0 {
+                    imp.progress.set(0.0);
+                    imp.phase.set((imp.phase.get() + 1) % 8);
+                } else {
+                    imp.progress.set(progress)
+                }
+                obj.queue_draw();
+            }));
+
+            let animation = adw::TimedAnimation::new(obj.as_ref(), 0.0, 1.0, 200, &target);
+            animation.set_easing(adw::Easing::EaseInOutQuad);
+
+            obj.connect_local("animate", true, move |_| {
+                let val = animation.value();
+                if val == 0.0 || val == 1.0 {
+                    animation.play()
+                }
+                None
+            });
+        }
+
+        fn properties() -> &'static [glib::ParamSpec] {
+            static PROPERTIES: Lazy<Vec<glib::ParamSpec>> = Lazy::new(|| {
+                vec![
+                    glib::ParamSpecBoxed::builder::<gdk::RGBA>("color1").build(),
+                    glib::ParamSpecBoxed::builder::<gdk::RGBA>("color2").build(),
+                    glib::ParamSpecBoxed::builder::<gdk::RGBA>("color3").build(),
+                    glib::ParamSpecBoxed::builder::<gdk::RGBA>("color4").build(),
+                    glib::ParamSpecBoolean::builder("dark").build(),
```

Comment:
We're probably gonna make this object settable via a tdlib object (like [this](https://docs.rs/tdlib/latest/tdlib/types/struct.Background.html)), so I don't think these properties are really something we need. I prefer having the colors hardcoded in this module for now, until we have a proper theme support (it will not happen in 0.1, too much code).

Reply:
I think those colors will be useful for settings preview